Sources/Document/CRDT/RHTPQMap.swift

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright 2022 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+/**
+ * `RHTPQMapNode` is a node of RHTPQMap.
+ */
+struct RHTPQMapNode: Equatable {
+    let rhtKey: String
+    let rhtValue: CRDTElement
+
+    fileprivate init(key: String, value: CRDTElement) {
+        self.rhtKey = key
+        self.rhtValue = value
+    }
+
+    /**
+     * `isRemoved` checks whether this value was removed.
+     */
+    func isRemoved() -> Bool {
+        return self.rhtValue.isRemoved()
+    }
+
+    /**
+     * `remove` removes a value base on removing time.
+     */
+    @discardableResult
+    fileprivate func remove(removedAt: TimeTicket) -> Bool {
+        return self.rhtValue.remove(removedAt)
+    }
+
+    static func == (lhs: RHTPQMapNode, rhs: RHTPQMapNode) -> Bool {
+        return lhs.rhtKey == rhs.rhtKey && lhs.rhtValue.equal(rhs.rhtValue)
+    }
+}
+
+/**
+ * RHTPQMap is replicated hash table with priority queue by creation time.
+ */
+class RHTPQMap {
+    private var elementQueueMapByKey: [String: Heap<TimeTicket, RHTPQMapNode>] = [:]
+    private var nodeMapByCreatedAt: [TimeTicket: RHTPQMapNode] = [:]
+
+    /**
+     * `set` sets the value of the given key.
+     */
+    @discardableResult
+    func set(key: String, value: CRDTElement) -> CRDTElement? {
+        var removed: CRDTElement?
+
+        if let queue = self.elementQueueMapByKey[key],
+           queue.length() >= 1,
```

Comment:
length()(count) may take longer than isEmpty. How about adding isEmpty variable at the Heap?

Reply:
Ok, `isEmpty` is better.